### ISSUE TITLE
Support For Low Latency HLS

### ISSUE THREAD

**Issue by SanatSharma** [videojs/http-streaming#721]
__Please do not delete the template, by filling out the required information we can investigate your issue more quickly.__

## Description
Apple recently released LL-HLS that allows super low-latency HLS. I was wondering when support for LL-HLS will be provided. They claim to have sub 2-sec latency, which would be great feature to have with (non-WebRTC) live streaming.

(https://developer.apple.com/documentation/http_live_streaming/protocol_extension_for_low-latency_hls_preliminary_specification)

**Comment by gkatsev** [comment]
Forgot to answer. We are aware of all the low latency HLS things but unfortunately we don't have the bandwidth (if you will) right now to work on it.

**Comment by jitendracodehacker** [comment]
@SanatSharma Did you get any breakthrough for this LHLS support for Video JS ? or what is your workaround?

**Comment by gkatsev** [comment]
This is something we are likely going to be working in Q3.

**Comment by SanatSharma** [comment]
@jitendracodehacker we ended up using webrtc

**Comment by Mariare2** [comment]
@gkatsev Have you been able to put any effort in it? Is there a planned release date on the LL HLS feature?

**Comment by gkatsev** [comment]
Unfortunately, we haven't been able to get started on this. We'll definitely make lots of noise when we get started on low latency streaming.

**Comment by heennkkee** [comment]
@SanatSharma Did you run WebRTC via video.js or did you end up with another player?

**Comment by danielehrhardt** [comment]
> @SanatSharma Did you run WebRTC via video.js or did you end up with another player?

New Status on this?

**Comment by gkatsev** [comment]
FYI, Video.js 7.13 has preliminary support for LHLS. If you try it out, please let us know how it is. Requires setting `experimentalLLHLS` on the player.

**Comment by n2quyet** [comment]
> 
> 
> FYI, Video.js 7.13 has preliminary support for LHLS. If you try it out, please let us know how it is. Requires setting `experimentalLLHLS` on the player.

I tried but it doesn't seem to work. the delay is about 20s, I also tried on theoplayer the delay is about 4s

**Comment by Phillipip** [comment]
Is there any working example? Unfortunately, it doesn't work for me in the [demo player](https://videojs-http-streaming.netlify.app/?debug=false&autoplay=false&muted=false&minified=false&sync-workers=false&liveui=false&llhls=true&url=https%3A%2F%2Fll-hls-test.apple.com%2Fmaster.m3u8&type=application%2Fx-mpegurl&keysystems=&buffer-water=false&exact-manifest-timings=false&pixel-diff-selector=false&override-native=true&preload=auto&mirror-source=false).

**Comment by gkatsev** [comment]
Can you elaborate on what you mean by "doesn't work"? It's still early and experimental. Definitely not ready for production use. I'm sure there's more work to do for it.

**Comment by Phillipip** [comment]
> Can you elaborate on what you mean by "doesn't work"?
I do not get LHLS, only HLS.

> It's still early and experimental.
Under which example URL can we successfully test this experiment?

> Definitely not ready for production use.
What is already working and what is not yet working?

**Comment by gkatsev** [comment]
It seems to me that parts are getting download when available.
Theoretically, it's all implemented, but it hasn't really gotten any testing and there's likely still bugs. We'll definitely make a big announcement when we think it's ready for people to use.

**Comment by Phillipip** [comment]
OK thank you!

**Comment by dovelive** [comment]
@gkatsev 
Hi.
I have test the Experimental LL HLS feature of VHS at VHS test site. https://videojs-http-streaming.netlify.app/
However compare to the THEOplayer(https://www.theoplayer.com/ll-hls-test-page), the Experimental LL HLS feature of VHS has some delay(about 5-8 seconds).
I was trying to change some config variables(e.g. GOAL_BUFFER_LENGTH, MAX_GOAL_BUFFER_LENGTH, etc...) in VHS, but there was no improvement.
So I realized that the delay between THEOPlayer and VHS on LL HLS Streaming Playing couldn't be adjusted at this time.
Would you like to make some progressive on decreasing the delay of LL HLS in VHS in the future.
Regards.

**Comment by gkatsev** [comment]
@dovelive this is actually something that we're working on. We have a PR out that should improve our starting time https://github.com/videojs/http-streaming/pull/1201 (feel free to try it out in the netlify link). In my testing, once we've started, we do a decent job and staying around the same time.
Once that PR is in, we can probably call LLHLS in beta. We do plan to eventually have some knobs that can be turned for LLHLS but it isn't available at this point. The current goal buffer config stuff don't really apply to LLHLS.
Also, thanks for trying out!

**Comment by dovelive** [comment]
@gkatsev 
Hi.
I have tested the http-streaming 2.12.0 on http-streaming test page.(https://videojs-http-streaming.netlify.app/)
Since the logic related to the playback of the LLHLS stream was completed than before, it was confirmed that the videojs player played the LLHLS streaming almost the same time as the THEOPlayer.(https://www.theoplayer.com/ll-hls-test-page)

However, I have one question.
In the THEOPlayer, user can move to any time by clicking any position in the Time Seekbar for the llhls stream.
However, for the videojs, it can't.
In other words, for the LLHLS stream, user cannot move to the desired time by clicking the on the SeekBar.
(I have tested it using https://ll-hls-test.apple.com/llhls1/multi.m3u8)

So would you like to check and implement this feature on the videojs too?

Regards.

**Comment by gkatsev** [comment]
@dovelive great to hear! For being able to seek in live streams, you need to enable `liveui` in Video.js. There's an option for it on https://videojs-http-streaming.netlify.app/ in the options tab panel. Worth noting that by default won't show it if the live window of the stream is less than 20 seconds in length because seeking in those streams is more likely to cause issues.
```js
videojs('video', {
  liveui: true
});
```

**Comment by dovelive** [comment]
@gkatsev 
Thank you for contacting.
Of course, I have tested LLHLS feature with liveui option checked.
But SeekToLive control not working for this stream. - https://ll-hls-test.apple.com/llhls1/multi.m3u8
And the action of clicking the on latest point of the seekbar does not work either.
Meanwhile THEOPlayer works fine for that stream too.(I mean SeekToLive button)

I am not sure there is a problem on that stream itself, however I want you to check it by yourself.

I hope my opinion would help for improving the performance of videojs player.
Regards.

![image](https://user-images.githubusercontent.com/19376598/143965934-c819548c-8db2-49f5-853a-4bb9deb23988.png)

**Comment by gkatsev** [comment]
This new apple stream is so far the only known issue for us. We haven't really had a chance to investigate further why this particular stream isn't working well. Other LLHLS streams that we tested work fine right now.

**Comment by gkatsev** [comment]
For those that haven't seen yet, there  is LLHLS support now.
It is flagged behind experimentalLLHLS. Ideally, you should also enable experimentalBufferBasedABR.

**Comment by ChR-iSz** [comment]
Do you habe more informations about `experimentalBufferBasedABR` ? Can't find any documentation for this...

**Comment by getroot** [comment]
![image](https://user-images.githubusercontent.com/17779085/171578821-476b8cd1-d4a2-4d9f-8b93-a593a3def631.png)

The LLHLS feature of videojs seems to work well with OvenMediaEngine as well. But the player doesn't seem to use #EXT-X-PRELOAD-HINT. Is this a server and player compatibility issue? Or is this not implemented in the player yet? (THEO player loads #EXT-X-PRELOAD-HINT)

I tested this with the URL below.

https://videojs-http-streaming.netlify.app/?debug=false&autoplay=false&muted=false&fluid=false&minified=false&sync-workers=false&liveui=true&llhls=true&url=https%3A%2F%2Fome-dev.airensoft.com%3A13334 %2Fapp%2Fstream%2Fllhls.m3u8&type=application%2Fx-mpegURL&keysystems=&buffer-water=true&exact-manifest-timings=true&pixel-diff-selector=true&network-info=false&dts=true-offset=false&override-native=true&preload=

For information about OvenMediaEngine's LLHLS, please refer to the URL below.

https://github.com/AirenSoft/OvenMediaEngine/issues/766

**Comment by doubledge-spec** [comment]
Isn't this implemented in vhs now?
